lib/ansible/module_utils/facts/virtual/linux.py

```diff
@@ -223,13 +223,15 @@ def get_virtual_facts(self):
         # In older Linux Kernel versions, /sys filesystem is not available
         # dmidecode is the safest option to parse virtualization related values
         dmi_bin = self.module.get_bin_path('dmidecode')
-        (rc, out, err) = self.module.run_command('%s -s system-product-name' % dmi_bin)
-        if rc == 0:
-            # Strip out commented lines (specific dmidecode output)
-            vendor_name = ''.join([line.strip() for line in out.splitlines() if not line.startswith('#')])
-            if vendor_name in ['VMware Virtual Platform', 'VMware7,1']:
-                virtual_facts['virtualization_type'] = 'VMware'
-                virtual_facts['virtualization_role'] = 'guest'
+        # We still want to continue even if dmidecode is not available
```

Comment:
Patch looks good as is. So :shipit: for me.

But was thinking it might be a good time to move this to a helper method, since then it could return early and lose a few indention levels if you wanted to.

``` python

def _collect_dmi(self):
    dmi_virt_facts = {}

    dmi_bin = self.module.get_bin_path('dmidecode')
    
    if not dmi_bin:
        return dmi_virt_facts
    
    (rc, out, err) = self.module.run_command('%s -s system-product-name' % dmi_bin)
    
    if rc != 0:
        return dmi_virt_facts

    # I might split this into two methods, one for finding and running dmi, and another for parsing it.
    # and passing in the dmidecode output to the parsing method
    # that would make the parsing code easy to test
    vendor_name = ''.join([line.strip() for line in out.splitlines() if not line.startswith('#')])
    
    if vendor_name in ['VMware Virtual Platform', 'VMware7,1']:
        dmi_virt_facts['virtualization_type'] = 'VMware'
        dmi_virt_facts['virtualization_role'] = 'guest'

    return dmi_virt_facts

```

with
``` python

dmi_virt_facts = self._collect_dmi()
virtual_facts.update(dmi_virt_facts)

```

or if split into two

``` python

dmi_bin_output = self._run_dmidecode()
dmi_virt_facts = self._collect_dmi(dmi_bin_output)
virtual_facts.update(dmi_virt_facts)

Reply:
I have one more suggestion, it would be good to replace

```
if vendor_name in ['VMware Virtual Platform', 'VMware7,1']:
```

with 

```
if vendor_name.startwith('VMware'):
```

I am not sure about `VMware 7,1` uniqueness on other ESXi version. (I missed this in last commit)

Reply:
@alikins IMHO that would only make sense if all the collections would follow the same logic in separate functions 
@Akasurde I agree and have updated the patch

Reply:
This introduced a regression, it's called start**s**with in Python.
Also this doesn't (and didn't) return early like all the other branches, so the VMware result would always be overwritten by NAs (see https://github.com/ansible/ansible/pull/34132#pullrequestreview-91756340).